pkg/provider/azure_utils.go

```diff
@@ -342,3 +341,152 @@ func extractVmssVMName(name string) (string, string, error) {
 	instanceID := split[len(split)-1]
 	return ssName, instanceID, nil
 }
+
+// getIPFamiliesEnabled checks if IPv4, IPv6 are enabled according to svc.Spec.IPFamilies.
+func getIPFamiliesEnabled(svc *v1.Service) (v4Enabled bool, v6Enabled bool) {
+	for _, ipFamily := range svc.Spec.IPFamilies {
+		if ipFamily == v1.IPv4Protocol {
+			v4Enabled = true
+		} else {
+			v6Enabled = true
+		}
+	}
+	return
+}
+
+// getServiceLoadBalancerIP retrieves LB IP from IPv4 annotation, then IPv6 annotation, then service.Spec.LoadBalancerIP.
+func getServiceLoadBalancerIP(service *v1.Service, isIPv6 bool) string {
+	if service == nil {
+		return ""
+	}
+
+	if ip, ok := service.Annotations[consts.ServiceAnnotationLoadBalancerIPDualStack[isIPv6]]; ok && ip != "" {
+		return ip
+	}
+
+	// Retrieve LB IP from service.Spec.LoadBalancerIP (will be deprecated)
+	svcLBIP := service.Spec.LoadBalancerIP
+	if (net.ParseIP(svcLBIP).To4() != nil && !isIPv6) ||
+		(net.ParseIP(svcLBIP).To4() == nil && isIPv6) {
+		return svcLBIP
+	}
+	return ""
+}
+
+func getServiceLoadBalancerIPs(service *v1.Service) []string {
+	if service == nil {
+		return []string{}
+	}
+
+	ips := []string{}
+	if ip, ok := service.Annotations[consts.ServiceAnnotationLoadBalancerIPDualStack[false]]; ok && ip != "" {
+		ips = append(ips, ip)
+	}
+	if ip, ok := service.Annotations[consts.ServiceAnnotationLoadBalancerIPDualStack[true]]; ok && ip != "" {
+		ips = append(ips, ip)
+	}
+	if len(ips) != 0 {
+		return ips
+	}
+
+	lbIP := service.Spec.LoadBalancerIP
+	if lbIP != "" {
+		ips = append(ips, lbIP)
+	}
+
+	return ips
+}
+
+// setServiceLoadBalancerIP sets LB IP to a Service
+func setServiceLoadBalancerIP(service *v1.Service, ip string) {
+	if service.Annotations == nil {
+		service.Annotations = map[string]string{}
+	}
+	isIPv6 := net.ParseIP(ip).To4() == nil
+	service.Annotations[consts.ServiceAnnotationLoadBalancerIPDualStack[isIPv6]] = ip
+}
+
+func getServicePIPName(service *v1.Service, isIPv6 bool) string {
+	if service == nil {
+		return ""
+	}
+
+	if consts.DualstackSupported {
```

Comment:
here is where I didn't get why a const is needed.

Reply:
Dualstack or not should be based on the service spec.

Reply:
Yes, I understand dualstack or not depends on Services' IPFamilies.
`DualstackSupported` here is a short-term feature gate. After all PRs have been merged, it will be deleted. Since this PR doesn't have all the feature code (will be a lot in azure_loadbalancer.go), without this feature gate, code won't work.

Reply:
I see. Could you add a TODO and say it would be removed in the next few PRs?

Reply:
Sure. There's also a feature gate in e2e code. Will add TODO here
https://github.com/kubernetes-sigs/cloud-provider-azure/blob/2230945407cc3f9490a4a54491f911870efdb009/tests/e2e/utils/network_utils.go#L50-L51